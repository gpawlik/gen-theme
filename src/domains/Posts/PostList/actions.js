import { createAction } from 'redux-actions';

import { actionTypes as at } from './constants';

export const fetch = createAction(
  at.POSTS_FETCH
);

export const fetchSuccess = createAction(
  at.POSTS_FETCH_SUCCESS, data => ({ data })
);

export const fetchError = createAction(
  at.POSTS_FETCH_ERROR, error => ({ error })
);

export const scrollDirection = createAction(
  at.POSTS_SCROLL, value => ({ value })
);

export const scrollUp = createAction(
  at.POSTS_SCROLL_UP
);

export const scrollDown = createAction(
  at.POSTS_SCROLL_DOWN
);
