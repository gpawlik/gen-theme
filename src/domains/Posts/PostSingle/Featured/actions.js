import { createAction } from 'redux-actions';

import { actionTypes as at } from './constants';

export const fetchFeaturedImage = createAction(
  at.POSTS_FEATURED_IMAGE_FETCH, imageId => ({ imageId })
);

export const fetchFeaturedImageSuccess = createAction(
  at.POSTS_FEATURED_IMAGE_FETCH_SUCCESS, data => ({ data })
);

export const fetchFeaturedImageError = createAction(
  at.POSTS_FEATURED_IMAGE_FETCH_ERROR, error => ({ error })
);
