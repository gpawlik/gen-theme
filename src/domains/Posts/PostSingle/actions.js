import { createAction } from 'redux-actions';

import { actionTypes as at } from './constants';

export const fetchSingle = createAction(
  at.POSTS_FETCH_SINGLE, postId => ({ postId })
);

export const fetchSingleSuccess = createAction(
  at.POSTS_FETCH_SINGLE_SUCCESS, data => ({ data })
);

export const fetchSingleError = createAction(
  at.POSTS_FETCH_SINGLE_ERROR, error => ({ error })
);
