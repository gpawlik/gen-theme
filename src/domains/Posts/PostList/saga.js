import { call, put, fork, takeLatest, throttle } from 'redux-saga/effects';

import { fetchPosts } from './service';
import { fetchSuccess, fetchError, scrollUp, scrollDown } from './actions';
import { actionTypes as at} from './constants';

export function* fetchAllPosts() {
  try {
    const result = yield call(fetchPosts);

    yield put(fetchSuccess(result));
  } catch (err) {
    yield put(fetchError(err));
  }
}

export function* scrollPosts({ payload }) {
  if (payload.value > 0) {
    yield put(scrollUp());
  } else {
    yield put(scrollDown());
  }
}

export function* postsWatcher() {
  yield fork(takeLatest, at.POSTS_FETCH, fetchAllPosts);
  yield throttle(0, at.POSTS_SCROLL, scrollPosts);
}
