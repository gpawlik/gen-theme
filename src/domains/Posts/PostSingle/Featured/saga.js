import { call, put, fork, takeLatest } from 'redux-saga/effects';

import { fetchFeatured } from './service';
import { fetchFeaturedImageSuccess, fetchFeaturedImageError } from './actions';
import { actionTypes as at} from './constants';

export function* fetchFeaturedImage({ payload }) {
  try {
    if (!payload.imageId) {
      throw new Error('Featured image Id missing.');
    }
    const result = yield call(fetchFeatured, payload.imageId);

    yield put(fetchFeaturedImageSuccess(result.source_url));
  } catch (err) {
    yield put(fetchFeaturedImageError(err));
  }
}

export function* featuredImageWatcher() {
  yield fork(takeLatest, at.POSTS_FEATURED_IMAGE_FETCH, fetchFeaturedImage);
}
