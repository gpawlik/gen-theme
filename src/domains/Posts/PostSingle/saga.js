import { call, put, fork, select, takeLatest } from 'redux-saga/effects';

import { fetchPost } from './service';
import { fetchSingleSuccess, fetchSingleError } from './actions';
import { actionTypes as at} from './constants';
import { selectPosts } from 'domains/Posts/PostList/selector';

export function* fetchSinglePost({ payload }) {
  try {
    const postId = parseInt(payload.postId, 10);
    const posts = yield(select(selectPosts()));
    const postFromStore = posts.find(item => item.id === postId);
    const result = postFromStore ? postFromStore : yield call(fetchPost, postId);

    yield put(fetchSingleSuccess(result));
  } catch (err) {
    yield put(fetchSingleError(err));
  }
}

export function* postWatcher() {
  yield fork(takeLatest, at.POSTS_FETCH_SINGLE, fetchSinglePost);
}
