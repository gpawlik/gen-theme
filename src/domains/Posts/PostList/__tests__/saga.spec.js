import { put } from 'redux-saga/effects';

import { fetchSuccess, fetchError } from '../actions';
import { fetchAllPosts } from '../saga';

const query = 'foo';

describe('fetchPosts Saga', () => {
  let fetchPostsGenerator;

  beforeEach(() => {
    fetchPostsGenerator = fetchAllPosts({ query });

    const callDescriptor = fetchPostsGenerator.next(query).value;

    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the fetchSuccess action if it requests the data successfully', () => {
    const response = [{
      title: 'foo'
    }];
    const putDescriptor = fetchPostsGenerator.next(response).value;

    expect(putDescriptor).toEqual(put(fetchSuccess(response, query)));
  });

  it('should call the fetchError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = fetchPostsGenerator.throw(response).value;

    expect(putDescriptor).toEqual(put(fetchError(response)));
  });
});
