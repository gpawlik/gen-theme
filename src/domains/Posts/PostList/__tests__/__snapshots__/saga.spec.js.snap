// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchPosts Saga should call the fetchError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchPosts Saga should dispatch the fetchSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;
